File: client-android/app/src/main/java/com/example/cloneapp/BackendClient.kt
package com.example.cloneapp

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.BufferedReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

class BackendClient(private val baseUrl: String = BuildConfig.BACKEND_URL) {
    suspend fun health(): String = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/health")
        val connection = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 5000
            readTimeout = 5000
        }
        return@withContext connection.use { conn ->
            val status = conn.responseCode
            val stream = if (status in 200..299) conn.inputStream else conn.errorStream
            val body = stream?.bufferedReader()?.use(BufferedReader::readText) ?: ""
            if (status == HttpURLConnection.HTTP_OK) {
                Regex("\"status\"\s*:\s*\"(.*?)\"").find(body)?.groupValues?.getOrNull(1) ?: "unknown"
            } else {
                "error"
            }
        }
    }

    suspend fun session(): SessionResponse = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/session")
        val connection = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 5000
            readTimeout = 5000
        }
        connection.use { conn ->
            val status = conn.responseCode
            val body = if (status in 200..299) {
                conn.inputStream.bufferedReader().use(BufferedReader::readText)
            } else {
                throw IllegalStateException("Session request failed with $status")
            }
            Json { ignoreUnknownKeys = true }.decodeFromString(SessionResponse.serializer(), body)
        }
    }

    companion object {
        fun default(context: Context): BackendClient {
            val url = context.getString(R.string.backend_url)
            return BackendClient(url)
        }
    }

    suspend fun login(email: String, password: String) = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/auth/login")
        val connection = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 5000
            readTimeout = 5000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
        }
        connection.use { conn ->
            OutputStreamWriter(conn.outputStream).use { writer ->
                writer.write(Json.encodeToString(mapOf("email" to email, "password" to password)))
            }
            val status = conn.responseCode
            if (status !in 200..299) throw IllegalStateException("Login failed with $status")
        }
    }

    suspend fun logout() = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/auth/logout")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 5000
            readTimeout = 5000
        }
        conn.use { connection ->
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Logout failed with $status")
        }
    }

    suspend fun advanceOnboarding(completed: Boolean, step: String? = null) = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/onboarding/advance")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 5000
            readTimeout = 5000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
        }
        conn.use { connection ->
            val payload = mutableMapOf("completed" to completed)
            if (step != null) payload["step"] = step
            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(Json.encodeToString(payload))
            }
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Advance onboarding failed with $status")
            val body = connection.inputStream.bufferedReader().use(BufferedReader::readText)
            Json.decodeFromString(OnboardingResponse.serializer(), body)
        }
    }

    suspend fun fetchFeed(): List<FeedItem> = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/feed")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 5000
            readTimeout = 5000
        }
        conn.use { connection ->
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Feed fetch failed with $status")
            val body = connection.inputStream.bufferedReader().use(BufferedReader::readText)
            Json.decodeFromString(FeedResponse.serializer(), body).items
        }
    }

    suspend fun contentDetail(contentId: String): ContentResponse = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/content/$contentId")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 5000
            readTimeout = 5000
        }
        conn.use { connection ->
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Content fetch failed with $status")
            val body = connection.inputStream.bufferedReader().use(BufferedReader::readText)
            Json.decodeFromString(ContentResponse.serializer(), body)
        }
    }

    suspend fun fetchPaywall(): PaywallResponse = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/paywall")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 5000
            readTimeout = 5000
        }
        conn.use { connection ->
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Paywall fetch failed with $status")
            val body = connection.inputStream.bufferedReader().use(BufferedReader::readText)
            Json.decodeFromString(PaywallResponse.serializer(), body)
        }
    }

    suspend fun purchase(planId: String) = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/paywall/purchase")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 5000
            readTimeout = 5000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
        }
        conn.use { connection ->
            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(Json.encodeToString(mapOf("planId" to planId)))
            }
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Purchase failed with $status")
        }
    }

    suspend fun notificationPreview(): NotificationPreview = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/notifications/preview")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 5000
            readTimeout = 5000
        }
        conn.use { connection ->
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Notification preview failed with $status")
            val body = connection.inputStream.bufferedReader().use(BufferedReader::readText)
            Json.decodeFromString(NotificationPreviewResponse.serializer(), body).notification
        }
    }

    suspend fun fetchProfile(): ProfileResponse = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/profile")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 5000
            readTimeout = 5000
        }
        conn.use { connection ->
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Profile fetch failed with $status")
            val body = connection.inputStream.bufferedReader().use(BufferedReader::readText)
            Json.decodeFromString(ProfileResponse.serializer(), body)
        }
    }

    suspend fun updateProfileName(name: String): ProfileResponse = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/profile")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 5000
            readTimeout = 5000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
        }
        conn.use { connection ->
            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(Json.encodeToString(mapOf("name" to name)))
            }
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Profile update failed with $status")
            val body = connection.inputStream.bufferedReader().use(BufferedReader::readText)
            Json.decodeFromString(ProfileResponse.serializer(), body)
        }
    }

    suspend fun updateFeatureFlags(paywallEnabled: Boolean? = null, notificationsEnabled: Boolean? = null): FeatureFlags = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/feature-flags")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 5000
            readTimeout = 5000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
        }
        conn.use { connection ->
            val payload = buildMap {
                paywallEnabled?.let { put("paywallEnabled", it) }
                notificationsEnabled?.let { put("notificationsEnabled", it) }
            }
            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(Json.encodeToString(payload))
            }
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Feature flag update failed with $status")
            val body = connection.inputStream.bufferedReader().use(BufferedReader::readText)
            Json.decodeFromString(FeatureFlagResponse.serializer(), body).featureFlags
        }
    }

    suspend fun updateProgress(contentId: String, position: Int, completed: Boolean): ContentProgress = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/content/$contentId/progress")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 5000
            readTimeout = 5000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
        }
        conn.use { connection ->
            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(Json.encodeToString(mapOf("position" to position, "completed" to completed)))
            }
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Progress update failed with $status")
            val body = connection.inputStream.bufferedReader().use(BufferedReader::readText)
            Json.decodeFromString(ProgressResponse.serializer(), body).progress
        }
    }

    suspend fun logEvent(name: String, properties: Map<String, String> = emptyMap()) = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/analytics/events")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 5000
            readTimeout = 5000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
        }
        conn.use { connection ->
            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(Json.encodeToString(mapOf("name" to name, "properties" to properties)))
            }
            connection.responseCode // ignore response
        }
    }
}

private inline fun <T> HttpURLConnection.use(block: (HttpURLConnection) -> T): T {
    return try {
        block(this)
    } finally {
        disconnect()
    }
}


    suspend fun requestDownload(contentId: String): DownloadResponse = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/content/$contentId/download")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 5000
            readTimeout = 5000
        }
        conn.use { connection ->
            val status = connection.responseCode
            if (status !in 200..299) throw IllegalStateException("Download request failed with $status")
            val body = connection.inputStream.bufferedReader().use(BufferedReader::readText)
            Json.decodeFromString(DownloadResponse.serializer(), body)
        }
    }

    suspend fun deleteDownload(contentId: String) = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/content/$contentId/download")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "DELETE"
            connectTimeout = 5000
            readTimeout = 5000
        }
        conn.use { connection ->
            connection.responseCode
        }
    }

File: client-ios/Sources/CloneUI/BackendClient.swift
import Foundation

public actor BackendClient {
    public enum ClientError: Error {
        case invalidURL
        case requestFailed
        case invalidResponse
    }

    public var baseURL: URL

    public init(baseURL: URL = URL(string: ProcessInfo.processInfo.environment["CLONE_BACKEND_URL"] ?? "http://localhost:4000")!) {
        self.baseURL = baseURL
    }

    public func health() async throws -> String {
        let url = baseURL.appending(path: "health")
        let (data, response) = try await URLSession.shared.data(from: url)
        guard
            let httpResponse = response as? HTTPURLResponse,
            httpResponse.statusCode == 200,
            let body = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let status = body["status"] as? String
        else {
            throw ClientError.requestFailed
        }
        return status
    }

    public func session() async throws -> SessionResponse {
        let url = baseURL.appending(path: "session")
        let (data, response) = try await URLSession.shared.data(from: url)
        guard let httpResponse = response as? HTTPURLResponse, (200..<300).contains(httpResponse.statusCode) else {
            throw ClientError.requestFailed
        }
        do {
            return try JSONDecoder().decode(SessionResponse.self, from: data)
        } catch {
            throw ClientError.invalidResponse
        }
    }

    public func login(email: String, password: String) async throws {
        var request = URLRequest(url: baseURL.appending(path: "auth/login"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(["email": email, "password": password])
        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
    }

    public func logout() async throws {
        var request = URLRequest(url: baseURL.appending(path: "auth/logout"))
        request.httpMethod = "POST"
        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
    }

    public func advanceOnboarding(completed: Bool, step: String? = nil) async throws -> SessionResponse.Onboarding {
        var request = URLRequest(url: baseURL.appending(path: "onboarding/advance"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        var payload: [String: Any] = ["completed": completed]
        if let step { payload["step"] = step }
        request.httpBody = try JSONSerialization.data(withJSONObject: payload, options: [])
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
        let decoded = try JSONDecoder().decode(OnboardingResponse.self, from: data)
        return decoded.onboarding
    }

    public func fetchFeed() async throws -> [FeedItem] {
        let url = baseURL.appending(path: "feed")
        let (data, response) = try await URLSession.shared.data(from: url)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
        return try JSONDecoder().decode(FeedResponse.self, from: data).items
    }

    public func contentDetail(id: String) async throws -> ContentResponse {
        let url = baseURL.appending(path: "content/\(id)")
        let (data, response) = try await URLSession.shared.data(from: url)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
        return try JSONDecoder().decode(ContentResponse.self, from: data)
    }

    public func updateProgress(id: String, position: Int, completed: Bool) async throws -> ContentProgress {
        var request = URLRequest(url: baseURL.appending(path: "content/\(id)/progress"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(["position": position, "completed": completed])
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
        return try JSONDecoder().decode(ProgressResponse.self, from: data).progress
    }

    public func fetchPaywall() async throws -> PaywallResponse {
        let url = baseURL.appending(path: "paywall")
        let (data, response) = try await URLSession.shared.data(from: url)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
        return try JSONDecoder().decode(PaywallResponse.self, from: data)
    }

    public func purchase(planId: String) async throws {
        var request = URLRequest(url: baseURL.appending(path: "paywall/purchase"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(["planId": planId])
        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
    }

    public func notificationPreview() async throws -> NotificationPreview {
        let url = baseURL.appending(path: "notifications/preview")
        let (data, response) = try await URLSession.shared.data(from: url)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
        return try JSONDecoder().decode(NotificationPreviewResponse.self, from: data).notification
    }

    public func fetchProfile() async throws -> ProfileResponse {
        let url = baseURL.appending(path: "profile")
        let (data, response) = try await URLSession.shared.data(from: url)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
        return try JSONDecoder().decode(ProfileResponse.self, from: data)
    }

    public func updateProfile(name: String) async throws -> SessionResponse.User {
        var request = URLRequest(url: baseURL.appending(path: "profile"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(["name": name])
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
        return try JSONDecoder().decode(ProfileResponse.self, from: data).user
    }

    public func updateFeatureFlags(paywallEnabled: Bool? = nil, notificationsEnabled: Bool? = nil) async throws -> SessionResponse.FeatureFlags {
        var request = URLRequest(url: baseURL.appending(path: "feature-flags"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        var payload: [String: Bool] = [:]
        if let paywallEnabled { payload["paywallEnabled"] = paywallEnabled }
        if let notificationsEnabled { payload["notificationsEnabled"] = notificationsEnabled }
        request.httpBody = try JSONEncoder().encode(payload)
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
        return try JSONDecoder().decode(FeatureFlagResponse.self, from: data).featureFlags
    }

    public func requestDownload(id: String) async throws -> DownloadResponse {
        var request = URLRequest(url: baseURL.appending(path: "content/\(id)/download"))
        request.httpMethod = "POST"
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
        return try JSONDecoder().decode(DownloadResponse.self, from: data)
    }

    public func deleteDownload(id: String) async throws {
        var request = URLRequest(url: baseURL.appending(path: "content/\(id)/download"))
        request.httpMethod = "DELETE"
        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw ClientError.requestFailed
        }
    }

    public func logEvent(name: String, properties: [String: String] = [:]) async {
        var request = URLRequest(url: baseURL.appending(path: "analytics/events"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONEncoder().encode(["name": name, "properties": properties])
        _ = try? await URLSession.shared.data(for: request)
    }
}

public struct SessionResponse: Decodable {
    public struct User: Decodable {
        public let authenticated: Bool
        public let name: String?
        public let avatarUrl: String?
        public let premium: Bool
    }

    public struct Onboarding: Decodable {
        public let completed: Bool
        public let step: String
    }

    public struct FeatureFlags: Decodable {
        public let paywallEnabled: Bool
        public let notificationsEnabled: Bool
    }

    public let status: String
    public let user: User
    public let onboarding: Onboarding
    public let featureFlags: FeatureFlags
    public let timestamp: String
}

private struct OnboardingResponse: Decodable {
    let onboarding: SessionResponse.Onboarding
}

public struct FeedResponse: Decodable {
    public let status: String
    public let items: [FeedItem]
}

public struct FeedItem: Decodable, Identifiable {
    public let id: String
    public let title: String
    public let subtitle: String
    public let imageUrl: String?
    public let duration: Int?
    public let difficulty: String?
    public let tags: [String]?
}

public struct ContentResponse: Decodable {
    public let status: String
    public let content: ContentDetail
    public let progress: ContentProgress
}

public struct ContentDetail: Decodable {
    public let id: String
    public let title: String
    public let description: String
    public let audioUrl: String
    public let duration: Int
    public let instructor: String
    public let coverUrl: String?
}

public struct ContentProgress: Decodable {
    public let position: Int
    public let completed: Bool
    public let updatedAt: String?
}

public struct PaywallResponse: Decodable {
    public let status: String
    public let paywallEnabled: Bool
    public let plans: [PaywallPlan]
    public let premium: Bool
}

public struct PaywallPlan: Decodable, Identifiable {
    public let id: String
    public let title: String
    public let priceLocalized: String
    public let trial: String?
    public let bestValue: Bool
}

public struct NotificationPreview: Decodable {
    public let title: String
    public let body: String
    public let deepLink: String
}

private struct NotificationPreviewResponse: Decodable {
    let status: String
    let notification: NotificationPreview
}

public struct ProfileResponse: Decodable {
    public let status: String
    public let user: SessionResponse.User
    public let featureFlags: SessionResponse.FeatureFlags
}

private struct FeatureFlagResponse: Decodable {
    let status: String
    let featureFlags: SessionResponse.FeatureFlags
}

private struct ProgressResponse: Decodable {
    let status: String
    let progress: ContentProgress
}

public struct DownloadResponse: Decodable {
    public let status: String
    public let download: DownloadInfo
}

public struct DownloadInfo: Decodable {
    public let id: String
    public let status: String
    public let downloadedAt: String
}
